### ISSUE TITLE
simp only [OfNat.ofNat] causes maximum recursion depth error

### ISSUE THREAD

**Issue by ammkrn** [leanprover/lean4#788]
### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your issue isn't already [filed](https://github.com/leanprover/lean4/issues).
    * Reduced the issue to a self-contained, reproducible test case.

### Description
Trying to unfold `OfNat.ofNat` using `simp` seems to pretty universally throw the error `maximum recursion depth has been reached (use `set_option maxRecDepth <num>` to increase limit)`. This even happens in pretty simple cases like with `(0 : Nat)`. I couldn't figure out why trying to unfold this is problematic, unless the literal `0` in `@OfNat.ofNat Nat 0 (instOfNatNat 0))` also unfolds to `OfNat.ofNat...`. As a side note, it would be convenient if errors of this kind told you what the current setting was (in this case, what the current `maxRecDepth` was). Thanks!

### Steps to Reproduce
```
example : (0 : Nat) = Nat.zero := by
  simp only [OfNat.ofNat]

example : (0 : Fin 9) = (Fin.ofNat 0) := by
  simp only [OfNat.ofNat]
```

**Expected behavior:** [What you expect to happen]
The expectation is that the definition of `OfNat.ofNat...` would be unfolded to whatever the particular instance definition is.

**Actual behavior:** [What actually happens]
maximum recursion depth error.

**Reproduces how often:** [What percentage of the time does it reproduce?]
Consistently

### Versions
Lean (version 4.0.0-nightly-2021-11-07, commit d5e05f31e4a7, Release)
MacOS 11.4

**Comment by leodemoura** [comment]
By unfolding an `OfNat.ofNat` application, a raw nat literal is produced. The same kind of raw nat literal produced by the `nat_lit <num>` notation. However, `simp` tries to convert raw `Nat` literals back to the `OfNat.ofNat` form, this step is currently hard-coded in the simplifier since it is only applied if the raw nat literal is not the child of an `OfNat.ofNat` application. These two steps together produce the infinite loop.

If you really need to unfold `OfNat.ofNat` in a proof, I suggest you use the tactic `delta OfNat.ofNat`.

Regarding the infinite loop, there are a few possible solutions:
- We produce an error if the user tries to unfold `OfNat.ofNat` using `simp`
- We disable the raw nat literal to `OfNat.ofNat` application step IF `OfNat.ofNat` is marked to be unfolded.
- We disable the raw nat literal to `OfNat.ofNat` application step IF the `only` modifier is used

**Comment by ammkrn** [comment]
@leodemoura

Thanks for clarifying the `OfNat.ofNat` simp process, I was able to get by with `delta` in the meantime. I ran into this with lemmas that involved a literal zero of some type (like `(0 : Fin n+1).val = ...`) where there was an easy defEq on the other side of the `OfNat` instance. Gabriel showed me one way of getting out of this specific situation by using `show` where the type annotation is the reduced form, like `show (Fin.ofNat 0).val = ... by simp [<definition of OfNat instance>]`, but that's not quite as obvious as delta or simp.

Options 2 and 3 seem like the most convenient from a user perspective, but they also seem like they might have a greater potential to cause auxiliary issues? Option 1 with a `try delta OfNat.ofNat` suggestion somewhere in the error might be a happy middle ground. In any case, thank you for the help.
